### PR TITLE
add hashmap size

### PR DESCRIPTION
**Pull request by PhoenixZeng**

### CODE REVIEW

File: wurst/data/HashMap.wurst [hunk 1]
```diff
@@ -29,6 +33,17 @@ public class HashMap<K,V> extends Table
 		remove(key)
 		return result
 
+	/** Remove all data from this map */
+	override function flush()
+		size = 0
+		super.flush()
+
+	function size()returns int
```
Comment (Frotty) [inline wurst/data/HashMap.wurst]: add space after )

File: wurst/data/HashMap.wurst [hunk 2]
```diff
@@ -29,6 +33,17 @@ public class HashMap<K,V> extends Table
...
+		return size
+
+	ondestroy
```
Comment (Frotty) [inline wurst/data/HashMap.wurst]: seems pointless? Shouldn't call anything on a destroyed instance.

File: wurst/data/HashMap.wurst [hunk 3]
```diff
@@ -104,9 +119,11 @@ public class IterableMap<K,V> extends HashMap<K, V>
 
 
 	/** Returns the length of this IterableMap **/
-	function size() returns int
+	override function size() returns int
 		return keys.size()
 
+	function mapSize() returns int
```
Comment (Frotty) [inline wurst/data/HashMap.wurst]: why is this here? size() and mapSize() should always be the same, right?

Comment (PhoenixZeng) [reply wurst/data/HashMap.wurst]: yes. and i use it in the hashmaptest.wurst to check the size . it doesn't make any sense. Do I need to get rid of it?

**Comment by PhoenixZeng** [comment]
`Few comments. I think there was an argument against adding size like this, because it adds "has" overhead to "put" and "remove". Idk, probably still fine tho.`
maybe i should make a sub class

File: wurst/data/HashMap.wurst (hunk 3 above)
Comment (Frotty) [reply wurst/data/HashMap.wurst]: Yes

File: wurst/data/HashMap.wurst (hunk 2 above)
Comment (Frotty) [reply wurst/data/HashMap.wurst]: Accessing values of destroyed instances is invalid

Comment (PhoenixZeng) [reply wurst/data/HashMap.wurst]: i already maked a new commit

**Comment by Frotty** [comment]
I'm not sure, I don't like having subclass just for size, but especially with typecasting the extra overhead could become important. Why not use `IterableMap` ? 🤔

**Comment by PhoenixZeng** [comment]
because the `IterableMap ` has a hashlist . That's more overhead then Sizedmap. 

for sizedlist:
The point is to know how many instances a class currently has. (Because map mapping some jass-object to class is common used in wurst.)
What can this bring?
1: For classes that do not need to destroy instances, only integer loops are needed to traverse instances, and linked list modules are not needed (the iteration overhead of linked list modules is higher).
2: easy for instances leaks check (especially for 8191 of 1.27)
3: If you use integers as keys, this is also the a implementation of `HashList` (I don't like `LinkedList` because it needs to create `LLEntry`)

**Comment by Frotty** [comment]
what is "sizedlist" ?

> For classes that do not need to destroy instances, only integer loops are needed to traverse instances, and linked list modules are not needed (the iteration overhead of linked list modules is higher)

unless you have nested iterations, you can save a LinkedListModule iterator and there won't be any allocation overhead.

> easy for instances leaks check (especially for 8191 of 1.27)

What does this have to do with maps? add a size var to the class, or make a module that does it for convenience.

> (I don't like LinkedList because it needs to create LLEntry)

Do you have an actual scenario where this is a problem?
LinkedLists perform best for most usecases. We don't want to split list structures up into too many different types (e.g. we had ArrayList, which might get a return) mostly because if they would all implement an interface, they would share id space. Otherwise you need to duplicate a lot of code for both types of list.

File: wurst/data/HashMap.wurst [hunk 4]
```diff
@@ -29,6 +29,28 @@ public class HashMap<K,V> extends Table
 		remove(key)
 		return result
 
+public class SizedMap<K,V> extends HashMap<K,V>
```
Comment (Frotty) [inline wurst/data/HashMap.wurst]: SizedMap kinda gives the illusion that the Map is constrained by a size, rather than just counting imo. @cokemonkey11 ?

Comment (Cokemonkey11) [reply wurst/data/HashMap.wurst]: I guess a size-constrained one might be called "SmallMap" or something. SizedMap feels to me like `Map<> implements Sized` so name feels ok

I don't see the point though in extending hashmap. Why not just add size to hashmap?

**Comment by PhoenixZeng** [comment]
`you can save a LinkedListModule iterator` 
yes .But that code would be more complicated than that. `for loop` is easy to use. and  Although nested iterations are used sparingly, they may also be used.
------------------------------------
idk. differert man have the differert uses for wurst. A subclass have only a size maybe not a good idea for the STD. This is a personal  class . I just need to add it to my library.

If there is no objection, we can close it.

File: wurst/data/HashMap.wurst (hunk 4 above)
Comment (Frotty) [reply wurst/data/HashMap.wurst]: guess the only argument is the introduced overhead.
If you think it doesn't matter, we should just add it to HashMap.

Comment (PhoenixZeng) [reply wurst/data/HashMap.wurst]: `guess the only argument is the introduced overhead.`
i think add more overhead is unnecessary (with only a feature )

Comment (Halithor) [reply wurst/data/HashMap.wurst]: Instead of adding this to the HashMap, why not just use the `IterableMap`? It already has the `size` method with minimal overhead (memory increase based on the number of keys). 

If we still want size for the a standard HashMap, I would prefer it goes into the standard HashMap. `size` methods are not uncommon methods for HashTable data structures and it would make sense for the Wurst implementation to have one.

Comment (PhoenixZeng) [reply wurst/data/HashMap.wurst]: because the `IterableMap` has a inner `hashlist` .
if u think `adding this to the HashMap` is good. i can do it

**Comment by Frotty** [comment]
just add it to hashmap then.

**Comment by Frotty** [comment]
can you fix the build?

**Comment by PhoenixZeng** [comment]
it fixd

File: wurst/data/HashMap.wurst [hunk 5]
```diff
@@ -7,12 +7,16 @@ import public Table
...
+	private int size = 0
+	
 	/** Whether a value exists under the given key or not */
 	function has(K key) returns boolean
 		return hasInt(key castTo int)
 
 	/** Saves the given value under the given key */
 	function put(K key, V value)
+		if not has(key)
+			size ++
```
Comment (Frotty) [inline wurst/data/HashMap.wurst]: Do not put spaces around unary operators

File: wurst/data/HashMap.wurst [hunk 6]
```diff
@@ -7,12 +7,16 @@ import public Table
 /** Generic Table Wrapper */
 public class HashMap<K,V> extends Table
 
+	private int size = 0
```
Comment (Frotty) [inline wurst/data/HashMap.wurst]: should probably be protected

File: wurst/data/HashMap.wurst [hunk 7]
```diff
@@ -29,6 +35,16 @@ public class HashMap<K,V> extends Table
...
 		return result
 
+	function size() returns int
+		return size
+
+	override function flush()
+		size = 0
+		super.flush()
+		
+	ondestroy
```
Comment (Frotty) [inline wurst/data/HashMap.wurst]: pointless, this already happens in Table

**Comment by Frotty** [comment]
few more things
